neat_ml/model/train.py

```diff
@@ -0,0 +1,343 @@
+    )
+
+    xgb = XGBClassifier(
+        objective="binary:logistic",
+        eval_metric="logloss",
+        n_estimators=300,
+        learning_rate=0.1,
+        max_depth=4,
+        subsample=0.8,
+        colsample_bytree=0.8,
+        scale_pos_weight=scale_pos_weight,
+        n_jobs=-1,
+        random_state=RANDOM_STATE,
+    )
+
+    ensemble = VotingClassifier(
+        estimators=[("rf", rf), ("xgb", xgb)],
+        voting="soft",
+        n_jobs=-1,
+    )
+
+    return Pipeline(
+        steps=[
+            ("impute", SimpleImputer(strategy="median")),
+            ("scale", StandardScaler()),
+            ("ensemble", ensemble),
+        ]
+    )
+
+
+def _scale_pos_weight(y: pd.Series) -> float:
+    """
+    Calculate the negative/positive class ratio 
+    for XGBoost's `scale_pos_weight`.
+
+    Parameters
+    ----------
+    y : pd.Series
+        The target variable Series containing 
+        binary class labels.
+
+    Returns
+    -------
+    float
+        The calculated ratio of negative to positive 
+        samples. Returns 1.0 if the positive class 
+        is not present.
+    """
+    counts = np.bincount(y, minlength=2)
+    neg, pos = counts
+    return neg / pos if pos else 1.0
+
+
+def train_with_validation(
+    X_train: pd.DataFrame,
+    y_train: pd.Series,
+    X_val: pd.DataFrame,
+    y_val: pd.Series,
+) -> Tuple[
+    Pipeline,
+    Dict[str, float],
+    Dict[str, int | float | None],
+    np.ndarray,
+]:
+    """
+    Train an ensemble classifier with hyperparameter 
+    tuning on a validation set.
+
+    This function iterates over a predefined 
+    hyperparameter grid, fitting each candidate model 
+    on the training set and evaluating its ROC-AUC 
+    score on the validation set. The parameters of 
+    the best model are then used to train a final model
+    on the combined training and validation data.
+
+    Parameters
+    ----------
+    X_train : pd.DataFrame
+        The feature matrix for the training set.
+    y_train : pd.Series
+        The target vector for the training set.
+    X_val : pd.DataFrame
+        The feature matrix for the validation set.
+    y_val : pd.Series
+        The target vector for the validation set.
+
+    Returns
+    -------
+    Tuple[Pipeline, Dict[str, float], 
+    Dict[str, int | float | None], np.ndarray]
+        A tuple containing:
+        - The final model pipeline, refit on 
+          the combined train+validation data.
+        - A dictionary of performance metrics 
+          (ROC-AUC, PR-AUC) on the validation set.
+        - A dictionary of the best hyperparameters
+          found during the grid search.
+        - The predicted probabilities for the positive
+          class on the validation set.
+    """
+    spw = _scale_pos_weight(y_train)
+    print(f"[INFO] scale_pos_weight={spw:.3f}  |  train neg/pos={np.bincount(y_train)}")
+
+    param_names = list(_PARAM_GRID.keys())
+    param_values = list(_PARAM_GRID.values())
```

Comment:
let's avoid using "module globals" like `_PARAM_GRID`, instead generally passing variables in through functions arguments to avoid spaghetti logic...